### DISCUSSION TITLE
Recursive type alias not working as expected

### DISCUSSION THREAD

**Discussion by guillaume-alliander** [microsoft/pyright#7557] [Q&A] [answered]
I'm pretty sure this is not a bug, that it is only me not understanding things properly, but there we are:

```python
import random
from typing import Union

type RecursiveType = Union[int, dict[int, "RecursiveType"]]
if random.randint(0, 1):
    m = 1
else:
    m = {1: 2}

rt: RecursiveType = m # Error here, see below
```

Pyright complains that `Expression of type "dict[int, int] | Literal[1]" cannot be assigned to declared type "RecursiveType"`

If I were to cast the `2` in the second part of the if to `RecursiveType`: `m = {1: cast(RecursiveType, 2)}` then pyright would be happy.

It feels to me as if Pyright should understand itself that the 2, being an `int`, is a valid `RecursiveType`.

I wonder what I do not understand?

(tested on the cli/vscode, version pyright-1.1.355)

**Comment by erictraut** [comment, marked as answer]
Pyright's behavior here is correct. The type parameters for the `dict` class are invariant, so the type `dict[int, int]` is not a subtype of `dict[int, RecursiveType]`. Not surprisingly, you will see the same error with mypy and other Python type checkers.

If you want this to type check, you will need to declare `m` as type `RecursiveType`. 

```python
m: RecursiveType
```

This will cause pyright to use bidirectional type inference when evaluating the type of the expression `{1: 2}`. Rather than evaluating it as `dict[int, int]`, which is what the normal dictionary inference rules produce, it will evaluate it as `dict[int, RecursiveType]`. Note that the type of an object is established at the time it is constructed. This dict is constructed at the point where the `{1: 2}` expression is evaluated. At the time of its "birth" it is either a `dict[int, int]` or a `dict[int, RecursiveType]`. It cannot be both. Nor can it change from one to the other after it is constructed.

For more details about type declarations, type inference, and bidirectional type inference, refer to [pyright's documentation](https://microsoft.github.io/pyright/#/type-concepts).

**Reply by guillaume-alliander** [reply]
Excellent, thanks for the detailed explanation!
